File: browser/src/segment.ts
// 动作帧均分定长分组 — 对应 Python core/segment.py。
// 把一个动作的去重帧按顺序每 groupSize 帧切一组 (默认 9 = 3×3 九宫格, 整批喂 AI)。
import type { Action, Cell } from './model';
import { uniqueCells } from './model';

export interface Segment {
  /** 组号 (从 1 起) */
  index: number;
  /** 组内首帧在 uniqueCells 里的序号 */
  start: number;
  /** 末帧序号 (含) */
  end: number;
  /** 该组的 (group, image) 帧坐标 */
  keys: Cell[];
}

/** 去重帧每 groupSize 切一组; 末组若 < 半组就并进前一组 (免落单一两帧)。
 *  mergeShortTail=false: 不并尾组 → 每组严格 ≤ groupSize (固定网格容量场景必须, 否则并入会超格)。 */
export function segmentAction(action: Action, groupSize = 9, mergeShortTail = true): Segment[] {
  const keys = uniqueCells(action);
  const gs = Math.max(1, groupSize);
  const starts: number[] = [];
  for (let s = 0; s < keys.length; s += gs) starts.push(s);
  if (mergeShortTail && starts.length >= 2 && keys.length - starts[starts.length - 1]! < gs / 2) {
    starts.pop(); // 末组太短 → 并进前一组
  }
  const segs: Segment[] = [];
  for (let n = 0; n < starts.length; n++) {
    const s = starts[n]!;
    const e = n + 1 < starts.length ? starts[n + 1]! : keys.length;
    segs.push({ index: n + 1, start: s, end: e - 1, keys: keys.slice(s, e) });
  }
  return segs;
}

File: core/segment.py
"""core/segment.py — 把一个动作的帧序列均分成固定大小的组(九宫格浏览 / 分批喂 AI)。

DNF 本体一个 IMG 塞了整套动作的全部帧。这里按顺序每 group_size 帧(默认 9, 正好
渲染成 3×3 九宫格)切一组 —— 均分、整齐, 每组帧少正好一批喂 nano banana。
(之前按姿势跳变的变长分段太碎, 改成均分定长。)
"""
from __future__ import annotations
from dataclasses import dataclass
from typing import List, Tuple

from core.model import SpriteSet, Action


@dataclass
class Segment:
    """一组(均分的一块)。"""
    index: int                       # 组号 (从 1 起)
    start: int                       # 组内首帧在 action.unique_cells() 里的序号
    end: int                         # 末帧序号 (含)
    keys: List[Tuple[int, int]]      # 该组的 (group, image) 帧坐标


def segment_action(sprites: SpriteSet, action: Action, group_size: int = 9) -> List[Segment]:
    """动作去重帧按顺序每 group_size 帧切一组 (默认 9 = 3×3 九宫格)。

    末组若太短 (< 半组) 就并进前一组, 免落单一两帧。sprites 参数保留以对齐调用契约。
    """
    keys = action.unique_cells()
    gs = max(1, group_size)
    starts = list(range(0, len(keys), gs))
    if len(starts) >= 2 and len(keys) - starts[-1] < gs / 2:   # 末组太短 → 并进前一组
        starts.pop()
    segs: List[Segment] = []
    for n, s in enumerate(starts):
        e = starts[n + 1] if n + 1 < len(starts) else len(keys)
        segs.append(Segment(index=n + 1, start=s, end=e - 1, keys=keys[s:e]))
    return segs
